src/model/level/internal.rs

```diff
@@ -0,0 +1,332 @@
+
+    #[serde(rename = "27", default, skip_serializing_if = "Option::is_none")]
+    pub password: Option<Internal<Password>>,
+
+    #[serde(rename = "28", default, skip_serializing_if = "Option::is_none")]
+    pub time_since_upload: Option<&'a str>,
+
+    #[serde(rename = "29", default, skip_serializing_if = "Option::is_none")]
+    pub time_since_update: Option<&'a str>,
+
+    #[serde(rename = "36", default, skip_serializing_if = "Option::is_none")]
+    pub index_36: Option<&'a str>,
+}
+
+impl<'a> HasRobtopFormat<'a> for Level<'a, Option<u64>, u64> {
+    type Internal = InternalLevel<'a>;
+
+    const DELIMITER: &'static str = ":";
+    const MAP_LIKE: bool = true;
+
+    fn as_internal(&'a self) -> Self::Internal {
+        InternalLevel {
+            level_id: self.level_id,
+            name: self.name.borrow(),
+            description: self.description.as_ref().map(|thunk| {
+                Internal(match thunk {
+                    Thunk::Unprocessed(unproc) => Thunk::Unprocessed(unproc),
+                    Thunk::Processed(Base64Decoded(moo)) => Thunk::Processed(Base64Decoded(Cow::Borrowed(moo.as_ref()))),
+                })
+            }),
+            version: self.version,
+            creator: self.creator,
+            is_auto: self.difficulty == LevelRating::Auto,
+            has_difficulty_rating: self.difficulty != LevelRating::NotAvailable,
+            rating: self.difficulty.into_response_value(),
+            is_demon: self.difficulty.is_demon(),
+            downloads: self.downloads,
+            main_song: self.main_song.unwrap_or(MAIN_SONGS[0]),
+            gd_version: self.gd_version.into(),
+            likes: self.likes,
+            length: self.length,
+            stars: self.stars,
+            featured: self.featured,
+            copy_of: self.copy_of,
+            index_31: self.index_31.as_ref().map(|moo| moo.borrow()),
+            custom_song: self.custom_song,
+            coin_amount: self.coin_amount,
+            coins_verified: self.coins_verified,
+            stars_requested: self.stars_requested,
+            index_40: self.index_40.as_ref().map(|moo| moo.borrow()),
+            is_epic: self.is_epic,
+            index_43: self.index_43.borrow(),
+            object_amount: self.object_amount,
+            index_46: self.index_46.as_ref().map(|moo| moo.borrow()),
+            index_47: self.index_47.as_ref().map(|moo| moo.borrow()),
+            level_data: self.level_data.as_ref().map(|data| data.level_data.borrow()),
+            password: self.level_data.as_ref().map(|data| Internal(data.password)),
+            time_since_upload: self.level_data.as_ref().map(|data| data.time_since_upload.borrow()),
+            time_since_update: self.level_data.as_ref().map(|data| data.time_since_update.borrow()),
+            index_36: self
+                .level_data
+                .as_ref()
+                .map(|data| data.index_36.as_ref().map(|moo| moo.borrow()))
```

Comment:
Is it maybe possible to swap .map and flatten here to reduce the code needed inside the map?

Reply:
We're dealing with an Option<LevelData> here, which in turn contains an `index_36: Option<&'a str>`, so we cannot flatten before the field index sadly. We can turn the first `.map` into `.and_then` and move the inner map outside though, if that's prettier :thinking: